### ISSUE TITLE
Can I use 2 more Server IP at extended configuration file format ?

### ISSUE THREAD

**Issue by obdurate0822** [shadowsocks/shadowsocks-rust#26]
Hi
**My server has 2 more public ip and use it by "alias ip"**

ex) 
eth0      Link encap:Ethernet  HWaddr 00:15:17:b6:5e:3c
          inet addr:**1.1.1.1**  Bcast:1.1.1.255  Mask:255.255.255.0
          inet6 addr: fe80::215:17ff:feb6:5e3c/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1667560 errors:0 dropped:0 overruns:0 frame:0
          TX packets:223040 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:642417843 (612.6 MiB)  TX bytes:25090036 (23.9 MiB)
          Interrupt:18 Memory:b8820000-b8840000

eth0:1    Link encap:Ethernet  HWaddr 00:15:17:b6:5e:3c
          inet addr:**1.1.1.2**  Bcast:1.1.1.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          Interrupt:18 Memory:b8820000-b8840000

each ip works well

so I congure shadowsocks-rust like this

{
    "servers": [
        {
            "address": "**1.1.1.1**",
            "port": 1080,
            "password": "hello-world",
            "method": "bf-cfb",
            "timeout": 300,
        },
        {
            "address": "**1.1.1.2**",
            "port": 1081,
            "password": "hello-kitty",
            "method": "aes-128-cfb"
        }
    ],
    "local_port": 8388,
    "local_address": "127.0.0.1"
}

**shadowsocks client connect to  1.1.1.1 ,1.1.1.2 without problem 
but after connection,If I check client's web public IP 
all IP check site only showing 1.1.1.1 even when I connect to 1.1.1.2

is there good solution for this situation ?**

**Comment by zonyitoo** [comment]
If I understand your description well, what you mean is:

You have configured shadowsocks with 2 servers, 1.1.1.1 and 1.1.1.2. You have verified that shadowsocks clients can connect to each of them. But when you use IP check site, even shadowsocks client have chosen different servers, they uses exactly the same public IP.

The `address` you have written in the configuration file is for shadowsocks to `bind`, which means that it only controls the LISTEN address.

Servers connect to remote websites with `TcpStream::connect`, it cannot choose which public IP it uses.

**Comment by obdurate0822** [comment]
thx for reply : )

I set 1 server with 2 ip (by alias ip eth0 1.1.1.1:1080 eth0:1 1.1.1.2:1081)
at client can connect to server by 1.1.1.1:1080 , 1.1.1.2:1081

after connection at 1.1.1.2:1081 
at client side,When I visit ip check site, the site show browser's ip is 1.1.1.1

ok then you mean,Can I bind Only 1 ip to shadowsocks server to Listen ?

**Comment by zonyitoo** [comment]
Huh? I can't get your point ...

First of all, shadowsocks' LISTEN IP doesn't equal to your public IP, they are not relevant. LISTEN IP is for accepting clients (shadowsocks client), and your public IP is for connecting to remote servers. shadowsocks can control the former case, but it cannot control the latter case.

**Comment by obdurate0822** [comment]
ok then..

like squid's "http_port" [http://www.squid-cache.org/Doc/config/http_port/](http://www.squid-cache.org/Doc/config/http_port/)

1.2.3.4:port
IP address with port
with shadowsocks can I serve multi server's public ip to each client ?
(may shadowsocks daemon at one machine or iptables... folk anything ok)

and what value can I set to config file's "server" item ? 
0.0.0.0 or 127.0.0.1 only ?

**Comment by zonyitoo** [comment]
The answer is: No.
This requires socket to bind one specific IP before `connect()`, and we don't have these APIs yet.

**Comment by Shawak** [comment]
@zonyitoo Will this be a thing in the future? I am having the same issue

**Comment by zonyitoo** [comment]
What problem have you encountered?

**Comment by Shawak** [comment]
I have two IPs on my first interface, now I want to use multiple servers and locals. The problem is that the outgoing IP from the servers is always my default IP, i can't bind it to another one.

My current basic test setup is the following:

```
{
    "server": "127.0.0.1",
    "server_port": 1080,
    "local_address": "<IP2>",
    "local_port": 8388,
    "password": "<password>",
    "timeout": 300,
    "method": "aes-128-gcm"
}
```
IP2 being my second IP on my default interface.

Now I start a sslocal and ssserver using this configuration file on my server and connect to the local (IP2:8838) using my browser. The expected behaviour is that it encrypts and decrypts the traffic internally and then routes the traffic through the interface using IP2, which results in me using IP2. But the ssserver uses IP1 as outgoing ip address.

This works using the libev version, but it throws bunch of errors in the console (for no reason) so I tried using your rust port of shadowsocks but encountered this problem.

**Comment by zonyitoo** [comment]
Well, what you want is

**Comment by Shawak** [comment]
Exactly, the libev version has the parameter -b (bind) for this purpose.

**Comment by zonyitoo** [comment]
I see.

1. Bind `local_address`: https://github.com/shadowsocks/shadowsocks-libev/blob/master/src/server.c#L478-L488
2. Bind interface: https://github.com/shadowsocks/shadowsocks-libev/blob/master/src/server.c#L490-L498

**Comment by zonyitoo** [comment]
@Shawak Please try if this commit work for you?

**Comment by zonyitoo** [comment]
This is a temporary solution. `tokio` haven't provide a direct API to call `bind` before `connect`. https://github.com/tokio-rs/tokio/issues/1543

**Comment by zonyitoo** [comment]
Should be fixed already. Merging to master.

**Comment by zonyitoo** [comment]
Released in v1.8.4 https://github.com/shadowsocks/shadowsocks-rust/releases/tag/v1.8.4
